<launch>
    <arg name="headless" default="true"/>
    <param name="robot_description" command="$(find xacro)/xacro '$(find hospitalbot)/urdf/hospitalbot.urdf.xacro'"/>
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node pkg="hospitalbot" type="odometry_publisher" name="odometry_publisher" output="screen"/>
    <node pkg="hospitalbot" type="key_publisher.py" name="keyboard_driver" output="screen"/>
    <node pkg="hospitalbot" type="command_twist_publisher" name="command_twist_publisher" output="screen">
        <param name="linear_velocity" value="0.25"/>
        <param name="angular_velocity" value="1.0"/>
    </node>
    <node pkg="rosserial_python" type="serial_node.py" name="serial_odometry_node" output="screen">
        <param name="port" value="/dev/odometry_arduino"/>
        <param name="baud" value="115200"/>
    </node>
    <node pkg="rosserial_python" type="serial_node.py" name="serial_imu_node" output="screen">
        <param name="port" value="/dev/imu_arduino"/>
        <param name="baud" value="115200"/>
    </node>
    <node name="rplidarNode" pkg="rplidar_ros" type="rplidarNode" output="screen">
        <param name="serial_port" type="string" value="/dev/ttyUSB0"/>  
        <param name="serial_baudrate" type="int" value="115200"/>
        <param name="frame_id" type="string" value="laser"/>
        <param name="inverted" type="bool" value="false"/>
        <param name="angle_compensate" type="bool" value="true"/>
        <remap from="scan" to="laser_scan"/>
    </node>
    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen">
        <remap from="scan" to="laser_scan"/>
        <!--param name="odom_frame" value="odom"/-->
        <param name="map_update_interval" value="2.0"/>
        <param name="maxUrange" value="6.0"/>
        <param name="sigma" value="0.05"/>
        <param name="kernelSize" value="1"/>
        <param name="lstep" value="0.05"/>
        <param name="astep" value="0.05"/>
        <param name="iterations" value="5"/>
        <param name="lsigma" value="0.075"/>
        <param name="ogain" value="3.0"/>
        <param name="lskip" value="0"/>
        <param name="srr" value="0.01"/>
        <param name="srt" value="0.02"/>
        <param name="str" value="0.01"/>
        <param name="stt" value="0.02"/>
        <param name="linearUpdate" value="0.25"/>     <!-- param name="linearUpdate" value="0.5"/-->
        <param name="angularUpdate" value="0.262"/>   <!--param name="angularUpdate" value="0.436"/-->
        <param name="temporalUpdate" value="-1.0"/>
        <param name="resampleThreshold" value="0.5"/>
        <param name="particles" value="300"/>
        <param name="xmin" value="-50.0"/>
        <param name="ymin" value="-50.0"/>
        <param name="xmax" value="50.0"/>
        <param name="ymax" value="50.0"/>
        <param name="delta" value="0.05"/>
        <param name="llsamplerange" value="0.01"/>
        <param name="llsamplestep" value="0.01"/>
        <param name="lasamplerange" value="0.005"/>
        <param name="lasamplestep" value="0.005"/>
    </node>
    <node pkg="laser_scan_matcher" type="laser_scan_matcher_node" name="laser_scan_matcher_node" output="screen">
        <remap from="scan" to="laser_scan"/>
        <remap from="imu/data" to="imu"/>
        <param name="fixed_frame" value = "odom"/>
        <param name="base_frame" value="base_footprint"/>
        <!-- <param name="use_alpha_beta" value="true"/> -->
        <param name="use_odom" value="false"/>
        <param name="use_imu" value="true"/>
        <param name="max_iterations" value="10"/>
        <param name="publish_pose" value="true"/>
        <param name="publish_tf" value="true"/>
        <param name="use_vel" value="false"/>
    </node>
    <node pkg="rviz" type="rviz" name="rviz" if="$(eval not headless)" args="-d $(find hospitalbot)/config/map.rviz"/>
</launch>